File: experiments/game-of-life/csharp/attempts/opus46high/3/GameOfLifeTests.cs
using System.Collections.Generic;
using Xunit;

namespace GameOfLife;

public class GameOfLifeTests
{
    private const Cell O = Cell.Dead;
    private const Cell X = Cell.Alive;
    private const Cell Z = Cell.Zombie;

    private static Grid MakeGrid(Cell[][] rows)
    {
        var height = rows.Length;
        var width = height == 0 ? 0 : rows[0].Length;
        var cells = new Dictionary<Position, Cell>();
        for (var row = 0; row < height; row++)
            for (var col = 0; col < width; col++)
                cells[new Position(row, col)] = rows[row][col];
        return new Grid(width, height, cells);
    }

    private static void AssertGridsEqual(Grid expected, Grid actual)
    {
        Assert.Equal(expected.Width, actual.Width);
        Assert.Equal(expected.Height, actual.Height);
        for (var row = 0; row < expected.Height; row++)
            for (var col = 0; col < expected.Width; col++)
            {
                var pos = new Position(row, col);
                Assert.Equal(expected.Cells[pos], actual.Cells[pos]);
            }
    }

    [Fact]
    public void Empty_matrix()
    {
        AssertGridsEqual(MakeGrid([]), MakeGrid([]).Tick());
    }

    [Fact]
    public void Live_cells_with_zero_live_neighbors_die()
    {
        var input = MakeGrid([[O, O, O], [O, X, O], [O, O, O]]);
        var expected = MakeGrid([[O, O, O], [O, O, O], [O, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Live_cells_with_only_one_live_neighbor_die()
    {
        var input = MakeGrid([[O, O, O], [O, X, O], [O, X, O]]);
        var expected = MakeGrid([[O, O, O], [O, O, O], [O, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Live_cells_with_two_live_neighbors_stay_alive()
    {
        var input = MakeGrid([[X, O, X], [X, O, X], [X, O, X]]);
        var expected = MakeGrid([[O, O, O], [X, O, X], [O, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Live_cells_with_three_live_neighbors_stay_alive()
    {
        var input = MakeGrid([[O, X, O], [X, O, O], [X, X, O]]);
        var expected = MakeGrid([[O, O, O], [X, O, O], [X, X, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Dead_cells_with_three_live_neighbors_become_alive()
    {
        var input = MakeGrid([[X, X, O], [O, O, O], [X, O, O]]);
        var expected = MakeGrid([[O, O, O], [X, X, O], [O, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Live_cells_with_four_or_more_neighbors_die()
    {
        var input = MakeGrid([[X, X, X], [X, X, X], [X, X, X]]);
        var expected = MakeGrid([[X, O, X], [O, O, O], [X, O, X]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Bigger_matrix()
    {
        var input = MakeGrid([
            [X, X, O, X, X, O, O, O],
            [X, O, X, X, O, O, O, O],
            [X, X, X, O, O, X, X, X],
            [O, O, O, O, O, X, X, O],
            [X, O, O, O, X, X, O, O],
            [X, X, O, O, O, X, X, X],
            [O, O, X, O, X, O, O, X],
            [X, O, O, O, O, O, X, X]
        ]);
        var expected = MakeGrid([
            [X, X, O, X, X, O, O, O],
            [O, O, O, O, O, X, X, O],
            [X, O, X, X, X, X, O, X],
            [X, O, O, O, O, O, O, X],
            [X, X, O, O, X, O, O, X],
            [X, X, O, X, O, O, O, X],
            [X, O, O, O, O, O, O, O],
            [O, O, O, O, O, O, X, X]
        ]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Zombie_with_zero_live_neighbors_stays_zombie()
    {
        var input = MakeGrid([[O, O, O], [O, Z, O], [O, O, O]]);
        var expected = MakeGrid([[O, O, O], [O, Z, O], [O, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Zombie_with_three_live_neighbors_stays_zombie()
    {
        var input = MakeGrid([[X, X, O], [O, Z, O], [X, O, O]]);
        var expected = MakeGrid([[X, X, O], [O, Z, O], [O, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Zombie_with_four_plus_live_neighbors_stays_zombie()
    {
        var input = MakeGrid([[X, X, X], [X, Z, X], [X, X, X]]);
        var expected = MakeGrid([[X, O, X], [O, Z, O], [X, O, X]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Dead_cell_with_three_zombie_neighbors_becomes_alive()
    {
        var input = MakeGrid([[Z, Z, O], [O, O, O], [Z, O, O]]);
        var expected = MakeGrid([[Z, Z, O], [X, X, O], [Z, O, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Live_cell_with_two_zombie_neighbors_stays_alive()
    {
        var input = MakeGrid([[Z, O, O], [O, X, O], [O, O, Z]]);
        var expected = MakeGrid([[Z, O, O], [O, X, O], [O, O, Z]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Live_cell_with_four_plus_zombie_neighbors_dies()
    {
        var input = MakeGrid([[Z, Z, O], [Z, X, O], [Z, O, O]]);
        var expected = MakeGrid([[Z, Z, O], [Z, O, O], [Z, X, O]]);
        AssertGridsEqual(expected, input.Tick());
    }

    [Fact]
    public void Mixed_grid_scenario()
    {
        var input = MakeGrid([
            [O, O, X, O, O, O, Z, O, O],
            [O, O, O, X, O, O, O, O, O],
            [O, X, X, X, O, O, O, O, O],
            [O, O, O, O, O, O, O, O, O],
            [Z, O, O, O, O, O, O, O, Z],
            [O, O, O, O, O, O, O, O, O],
            [O, O, O, O, O, X, X, X, O],
            [O, O, O, O, O, X, O, O, O],
            [O, O, Z, O, O, X, O, Z, O]
        ]);
        var expected = MakeGrid([
            [O, O, O, O, O, O, Z, O, O],
            [O, X, O, X, O, O, O, O, O],
            [O, O, X, X, O, O, O, O, O],
            [O, X, X, O, O, O, O, O, O],
            [Z, O, O, O, O, O, O, O, Z],
            [O, O, O, O, O, O, X, X, O],
            [O, O, O, O, O, X, X, O, O],
            [O, O, O, O, X, X, O, X, O],
            [O, O, Z, O, O, O, X, Z, O]
        ]);
        AssertGridsEqual(expected, input.Tick());
    }
}

File: experiments/game-of-life/rustlang/attempts/opus46high/5/tests/game_of_life_tests.rs
#[cfg(test)]
mod tests {
    use game_of_life::*;
    use std::collections::HashMap;

    const O: Cell = Cell::Dead;
    const X: Cell = Cell::Alive;
    const Z: Cell = Cell::Zombie;

    fn grid(rows: &[&[Cell]]) -> Grid {
        let height = rows.len() as i32;
        let width = if height == 0 { 0 } else { rows[0].len() as i32 };

        let mut cells = HashMap::new();
        for (r, row) in rows.iter().enumerate() {
            for (c, &cell) in row.iter().enumerate() {
                cells.insert(
                    Position {
                        row: r as i32,
                        col: c as i32,
                    },
                    cell,
                );
            }
        }

        Grid {
            width,
            height,
            cells,
        }
    }

    #[test]
    fn empty_matrix() {
        assert_eq!(tick_grid(&grid(&[])), grid(&[]));
    }

    #[test]
    fn live_cells_with_zero_live_neighbors_die() {
        let input = grid(&[
            &[O, O, O],
            &[O, X, O],
            &[O, O, O],
        ]);
        let expected = grid(&[
            &[O, O, O],
            &[O, O, O],
            &[O, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn live_cells_with_only_one_live_neighbor_die() {
        let input = grid(&[
            &[O, O, O],
            &[O, X, O],
            &[O, X, O],
        ]);
        let expected = grid(&[
            &[O, O, O],
            &[O, O, O],
            &[O, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn live_cells_with_two_live_neighbors_stay_alive() {
        let input = grid(&[
            &[X, O, X],
            &[X, O, X],
            &[X, O, X],
        ]);
        let expected = grid(&[
            &[O, O, O],
            &[X, O, X],
            &[O, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn live_cells_with_three_live_neighbors_stay_alive() {
        let input = grid(&[
            &[O, X, O],
            &[X, O, O],
            &[X, X, O],
        ]);
        let expected = grid(&[
            &[O, O, O],
            &[X, O, O],
            &[X, X, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn dead_cells_with_three_live_neighbors_become_alive() {
        let input = grid(&[
            &[X, X, O],
            &[O, O, O],
            &[X, O, O],
        ]);
        let expected = grid(&[
            &[O, O, O],
            &[X, X, O],
            &[O, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn live_cells_with_four_or_more_neighbors_die() {
        let input = grid(&[
            &[X, X, X],
            &[X, X, X],
            &[X, X, X],
        ]);
        let expected = grid(&[
            &[X, O, X],
            &[O, O, O],
            &[X, O, X],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn bigger_matrix() {
        let input = grid(&[
            &[X, X, O, X, X, O, O, O],
            &[X, O, X, X, O, O, O, O],
            &[X, X, X, O, O, X, X, X],
            &[O, O, O, O, O, X, X, O],
            &[X, O, O, O, X, X, O, O],
            &[X, X, O, O, O, X, X, X],
            &[O, O, X, O, X, O, O, X],
            &[X, O, O, O, O, O, X, X],
        ]);
        let expected = grid(&[
            &[X, X, O, X, X, O, O, O],
            &[O, O, O, O, O, X, X, O],
            &[X, O, X, X, X, X, O, X],
            &[X, O, O, O, O, O, O, X],
            &[X, X, O, O, X, O, O, X],
            &[X, X, O, X, O, O, O, X],
            &[X, O, O, O, O, O, O, O],
            &[O, O, O, O, O, O, X, X],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn zombie_with_zero_live_neighbors_stays_zombie() {
        let input = grid(&[
            &[O, O, O],
            &[O, Z, O],
            &[O, O, O],
        ]);
        let expected = grid(&[
            &[O, O, O],
            &[O, Z, O],
            &[O, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn zombie_with_three_live_neighbors_stays_zombie() {
        let input = grid(&[
            &[X, X, O],
            &[O, Z, O],
            &[X, O, O],
        ]);
        let expected = grid(&[
            &[X, X, O],
            &[O, Z, O],
            &[O, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn zombie_with_four_plus_live_neighbors_stays_zombie() {
        let input = grid(&[
            &[X, X, X],
            &[X, Z, X],
            &[X, X, X],
        ]);
        let expected = grid(&[
            &[X, O, X],
            &[O, Z, O],
            &[X, O, X],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn dead_cell_with_three_zombie_neighbors_becomes_alive() {
        let input = grid(&[
            &[Z, Z, O],
            &[O, O, O],
            &[Z, O, O],
        ]);
        let expected = grid(&[
            &[Z, Z, O],
            &[X, X, O],
            &[Z, O, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn live_cell_with_two_zombie_neighbors_stays_alive() {
        let input = grid(&[
            &[Z, O, O],
            &[O, X, O],
            &[O, O, Z],
        ]);
        let expected = grid(&[
            &[Z, O, O],
            &[O, X, O],
            &[O, O, Z],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn live_cell_with_four_plus_zombie_neighbors_dies() {
        let input = grid(&[
            &[Z, Z, O],
            &[Z, X, Z],
            &[O, O, O],
        ]);
        let expected = grid(&[
            &[Z, Z, X],
            &[Z, O, Z],
            &[O, X, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }

    #[test]
    fn mixed_grid_integration() {
        let input = grid(&[
            &[O, O, X, O, O, O, Z, O, O],
            &[O, O, O, X, O, O, O, O, O],
            &[O, X, X, X, O, O, O, O, O],
            &[O, O, O, O, O, O, O, O, O],
            &[Z, O, O, O, O, O, O, O, Z],
            &[O, O, O, O, O, O, O, O, O],
            &[O, O, O, O, O, X, X, X, O],
            &[O, O, O, O, O, X, O, O, O],
            &[O, O, Z, O, O, X, O, Z, O],
        ]);
        let expected = grid(&[
            &[O, O, O, O, O, O, Z, O, O],
            &[O, X, O, X, O, O, O, O, O],
            &[O, O, X, X, O, O, O, O, O],
            &[O, X, X, O, O, O, O, O, O],
            &[Z, O, O, O, O, O, O, O, Z],
            &[O, O, O, O, O, O, X, X, O],
            &[O, O, O, O, O, X, X, O, O],
            &[O, O, O, O, X, X, O, X, O],
            &[O, O, Z, O, O, O, X, Z, O],
        ]);
        assert_eq!(tick_grid(&input), expected);
    }
}
